File: runtime/ts/src/identity_wire.test.ts
import assert from 'node:assert/strict';
import test, { type TestContext } from 'node:test';
import type { Endpoint } from '@nightseam/duplex';
import { IDENTITY_METHOD, identityHandler } from './identity.ts';
import { prepareIdentity } from './identity_wire.ts';
import { callWire, handleWire, emitWire, onWireEvent } from './wire.ts';
import { wirePair } from './wire-pair.ts';
import { createDispatcher } from './dispatcher.ts';

const expected = { path: 'service', digest: 'a'.repeat(64) };
function deferred<T = void>() {
  let resolve!: (value: T) => void;
  const promise = new Promise<T>((yes) => {
    resolve = yes;
  });
  return { promise, resolve };
}
function dispatchedPair(t: TestContext, options: Parameters<typeof wirePair>[0] = {}) {
  const [near, far] = wirePair(options);
  const nearDispatcher = createDispatcher(near),
    farDispatcher = createDispatcher(far);
  t.after(() => near.close());
  t.after(() => nearDispatcher.close());
  t.after(() => farDispatcher.close());
  return { far, nearDispatcher, farDispatcher };
}
function observing(wire: Endpoint, observe: (kind: string) => void): Endpoint {
  return {
    send: (path, message) => wire.send(path, message),
    close: (code, reason) => wire.close(code, reason),
    receive(receiver) {
      return wire.receive({
        ...receiver,
        message(path, message) {
          observe(message.frame.kind);
          return receiver.message?.(path, message);
        },
      });
    },
  };
}

for (const absent of [false, true])
  test(`identity preparation retains the earliest event through check and binding (${absent ? 'absent' : 'matching'})`, async (t) => {
    const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
    if (!absent) handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
    const arrived = deferred(),
      delivered = deferred<number>();
    const preparation = prepareIdentity(
      observing(nearDispatcher.select([]), () => arrived.resolve()),
      expected,
    );
    t.after(() => preparation.close());
    let effects = 0;
    onWireEvent(preparation.wire, ['event'], (value) => {
      effects++;
      delivered.resolve(value as number);
    });
    emitWire(far, ['event'], 7);
    await arrived.promise;
    await preparation.check();
    assert.equal(effects, 0, 'checked but not yet bound');
    preparation.ready();
    assert.equal(await delivered.promise, 7);
    assert.equal(effects, 1);
  });

test('identity mismatch discards a held event and preserves other users of the carrier', async (t) => {
  const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler({ ...expected, digest: 'b'.repeat(64) }));
  const arrived = deferred();
  const preparation = prepareIdentity(
    observing(nearDispatcher.select([]), () => arrived.resolve()),
    expected,
  );
  t.after(() => preparation.close());
  let effects = 0;
  onWireEvent(preparation.wire, ['event'], () => {
    effects++;
  });
  handleWire(nearDispatcher, ['unrelated'], () => 19);
  emitWire(far, ['event'], 7);
  await arrived.promise;
  await assert.rejects(preparation.check(), { code: 'contract_mismatch' });
  assert.throws(() => preparation.ready(), { code: 'contract_mismatch' });
  assert.equal(await callWire(far, ['unrelated']), 19);
  assert.equal(effects, 0);
  assert.doesNotThrow(() => onWireEvent(nearDispatcher, ['event'], () => {}));
});

test('deferred request does not block the root and cancellation never dispatches later', async (t) => {
  const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
  const arrived = deferred(),
    cancelled = deferred();
  const preparation = prepareIdentity(
    observing(nearDispatcher.select([]), (kind) => {
      if (kind === 'request') arrived.resolve();
      if (kind === 'cancel') cancelled.resolve();
    }),
    expected,
  );
  t.after(() => preparation.close());
  let effects = 0;
  handleWire(preparation.wire, ['model'], () => {
    effects++;
    return 11;
  });
  handleWire(nearDispatcher, ['unrelated'], () => 7);
  const controller = new AbortController();
  const first = callWire(far, ['model'], null, { signal: controller.signal }).catch((error: unknown) => error);
  await arrived.promise;
  assert.equal(await callWire(far, ['unrelated']), 7);
  controller.abort();
  assert.equal(((await first) as { code: string }).code, 'cancelled');
  await cancelled.promise;
  await preparation.check();
  preparation.ready();
  assert.equal(await callWire(far, ['model']), 11);
  assert.equal(effects, 1);
});

for (const checked of [false, true])
  test(`preparation bounds a factory ${checked ? 'never bound' : 'never started'}`, async (t) => {
    const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
    handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
    const preparation = prepareIdentity(nearDispatcher.select([]), expected, { requestTimeoutMs: 250 });
    t.after(() => preparation.close());
    handleWire(preparation.wire, ['model'], () => {
      throw new Error('expired model dispatched');
    });
    if (checked) await preparation.check();
    await assert.rejects(callWire(far, ['model']), { code: 'cancelled' });
    assert.throws(() => preparation.ready(), { code: 'cancelled' });
    assert.doesNotThrow(() => handleWire(nearDispatcher, ['model'], () => 1));
    const replacement = prepareIdentity(nearDispatcher.select([]), expected);
    replacement.close();
  });

test('cancellation and explicit close abandon only their interpretation', async (t) => {
  const { far, nearDispatcher } = dispatchedPair(t);
  const preparation = prepareIdentity(nearDispatcher.select([]), expected);
  t.after(() => preparation.close());
  handleWire(preparation.wire, ['model'], () => {
    throw new Error('cancelled model dispatched');
  });
  handleWire(nearDispatcher, ['unrelated'], () => 7);
  await assert.rejects(preparation.check({ signal: AbortSignal.abort() }), { code: 'cancelled' });
  assert.equal(await callWire(far, ['unrelated']), 7);
  const second = prepareIdentity(nearDispatcher.select([]), expected);
  handleWire(second.wire, ['model'], () => 1);
  second.close();
  assert.throws(() => second.ready(), { code: 'disconnected' });
  assert.equal(await callWire(far, ['unrelated']), 7);
});

test('carrier closure releases held events and model requests', async (t) => {
  const { far, nearDispatcher } = dispatchedPair(t);
  const arrived = deferred(),
    closed = deferred();
  const preparation = prepareIdentity(
    observing(nearDispatcher.select([]), () => arrived.resolve()),
    expected,
  );

  preparation.wire.register(['event'], {
    message: () => {
      throw new Error('closed model dispatched');
    },
    closed: () => closed.resolve(),
  });
  emitWire(far, ['event']);
  await arrived.promise;
  far.close();
  await closed.promise;
  assert.throws(() => preparation.ready(), { code: 'disconnected' });
});

test('deferred request budget is bounded separately from the carrier budget', async (t) => {
  const { far, nearDispatcher, farDispatcher } = dispatchedPair(t, { maxConcurrentHandlers: 8 });
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
  const arrived = deferred();
  const preparation = prepareIdentity(
    observing(nearDispatcher.select([]), () => arrived.resolve()),
    expected,
    { maxConcurrentHandlers: 1 },
  );
  t.after(() => preparation.close());
  handleWire(preparation.wire, ['model'], () => 11);
  const first = callWire(far, ['model']);
  await arrived.promise;
  await assert.rejects(callWire(far, ['model']), { code: 'busy' });
  await preparation.check();
  preparation.ready();
  assert.equal(await first, 11);
});

test('readiness cannot precede the check, and check/ready are once only', async (t) => {
  const { nearDispatcher, farDispatcher } = dispatchedPair(t);
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
  const preparation = prepareIdentity(nearDispatcher.select([]), expected);
  t.after(() => preparation.close());
  assert.throws(() => preparation.ready(), /not been checked/);
  await assert.rejects(callWire(preparation.wire, ['model']), { code: 'busy' });
  await preparation.check();
  await assert.rejects(preparation.check(), /already started/);
  preparation.ready();
  assert.throws(() => preparation.ready(), /already ready/);
});

test('identity mismatch refuses an already deferred model request', async (t) => {
  const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler({ ...expected, digest: 'b'.repeat(64) }));
  const arrived = deferred();
  const preparation = prepareIdentity(
    observing(nearDispatcher.select([]), () => arrived.resolve()),
    expected,
  );
  t.after(() => preparation.close());
  handleWire(preparation.wire, ['model'], () => {
    throw new Error('mismatched request dispatched');
  });
  const pending = callWire(far, ['model']).catch((error: unknown) => error);
  await arrived.promise;
  await assert.rejects(preparation.check(), { code: 'contract_mismatch' });
  assert.equal(((await pending) as { code: string }).code, 'contract_mismatch');
});

test('closing immediately after ready answers each deferred request only once', async (t) => {
  const { far, nearDispatcher, farDispatcher } = dispatchedPair(t);
  handleWire(farDispatcher, [IDENTITY_METHOD], identityHandler(expected));
  const arrived = deferred();
  let responseCount = () => 0;
  const selected = nearDispatcher.select([]);
  const source: Endpoint = {
    send: (path, message) => selected.send(path, message),
    close: (code, reason) => selected.close(code, reason),
    receive(receiver) {
      return selected.receive({
        ...receiver,
        message(path, message) {
          if (path[0] === 'model' && message.frame.kind === 'request' && message.return) {
            // Spy in place, retaining this runtime-created Message and return
            // capability with their existing private invocation association.
            const sending = t.mock.method(message.return.wire, 'send');
            // The outcome is what this counts: the same capability also
            // carries the invocation's own lifecycle vocabulary at its other
            // paths, which is not an answer to the request.
            responseCount = () =>
              sending.mock.calls.filter((call) => (call.arguments[0] as readonly string[]).length === 0).length;
            arrived.resolve();
          }
          return receiver.message?.(path, message);
        },
      });
    },
  };
  const preparation = prepareIdentity(source, expected);
  handleWire(preparation.wire, ['model'], () => {
    throw new Error('closed model dispatched');
  });
  const pending = callWire(far, ['model']).catch((error: unknown) => error);
  await arrived.promise;
  await preparation.check();
  preparation.ready();
  preparation.close();
  assert.equal(((await pending) as { code: string }).code, 'disconnected');
  await new Promise<void>((resolve) => setImmediate(resolve));
  assert.equal(responseCount(), 1);
});

File: runtime/go/identity_wire_test.go
package runtime

import (
	"context"
	"encoding/json"
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"strings"
	"sync/atomic"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
)

var preparedIdentity = DeclarationIdentity{Path: "service", Digest: strings.Repeat("a", 64)}

type observedIdentityWire struct {
	bitwire.Endpoint
	observed chan bitwire.ProfileKind
}

func (w observedIdentityWire) Receive(receiver bitwire.Receiver) (func(), error) {
	original := receiver.Message
	receiver.Message = func(path []string, message bitwire.Message) {
		w.observed <- message.Frame.Kind
		original(path, message)
	}
	return w.Endpoint.Receive(receiver)
}
func identityWait(t *testing.T, signal <-chan bitwire.ProfileKind, kind bitwire.ProfileKind) {
	t.Helper()
	select {
	case actual := <-signal:
		if actual != kind {
			t.Fatalf("delivery %s, want %s", actual, kind)
		}
	case <-time.After(5 * time.Second):
		t.Fatal("delivery did not arrive")
	}
}
func installWireIdentity(t *testing.T, wire bitwire.Endpoint, identity DeclarationIdentity) {
	t.Helper()
	handler, err := IdentityHandler(identity)
	if err != nil {
		t.Fatal(err)
	}
	_, err = HandleWire(testBinding(t, wire), []string{IdentityMethod}, func(ctx context.Context, raw json.RawMessage) (any, error) { return handler(ctx, nil, raw) })
	if err != nil {
		t.Fatal(err)
	}
}
func requireIdentityCode(t *testing.T, err error, code string) {
	t.Helper()
	var public *PublicError
	if !errors.As(err, &public) || public.Code != code {
		t.Fatalf("error %v, want %s", err, code)
	}
}

func TestIdentityPreparationHoldsFirstEventThroughCheckAndBinding(t *testing.T) {
	for _, absent := range []bool{false, true} {
		t.Run(map[bool]string{false: "matching", true: "absent"}[absent], func(t *testing.T) {
			near, far := localPair(t, Options{})
			if !absent {
				installWireIdentity(t, far, preparedIdentity)
			}
			observed := make(chan bitwire.ProfileKind, 8)
			p, err := PrepareIdentity(observedIdentityWire{near, observed}, preparedIdentity, Options{})
			if err != nil {
				t.Fatal(err)
			}
			defer p.Close()
			delivered := make(chan int, 2)
			_, err = RegisterWire(p.Wire(), []string{"event"}, WireHandlers{Event: func(_ context.Context, raw json.RawMessage) error {
				var value int
				_ = json.Unmarshal(raw, &value)
				delivered <- value
				return nil
			}})
			if err != nil {
				t.Fatal(err)
			}
			if err = EmitWire(context.Background(), far, []string{"event"}, 1); err != nil {
				t.Fatal(err)
			}
			identityWait(t, observed, bitwire.ProfileEvent)
			// The local root's event delivery is held. Identity's response returns
			// through its original return address, never through that event FIFO.
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			if err = p.Check(ctx); err != nil {
				t.Fatal(err)
			}
			select {
			case <-delivered:
				t.Fatal("model ran before binding")
			default:
			}
			if err = p.Ready(); err != nil {
				t.Fatal(err)
			}
			select {
			case value := <-delivered:
				if value != 1 {
					t.Fatal(value)
				}
			case <-ctx.Done():
				t.Fatal("first event lost")
			}
		})
	}
}

func TestIdentityPreparationMismatchDiscardsEventsAndPreservesSharedCarrier(t *testing.T) {
	near, far := localPair(t, Options{})
	installWireIdentity(t, far, DeclarationIdentity{Path: "service", Digest: strings.Repeat("b", 64)})
	root := testBinding(t, near)
	model := root.Select(nil)
	observed := make(chan bitwire.ProfileKind, 8)
	p, err := PrepareIdentity(observedIdentityWire{model, observed}, preparedIdentity, Options{})
	if err != nil {
		t.Fatal(err)
	}
	defer p.Close()
	var effects atomic.Int32
	_, _ = RegisterWire(p.Wire(), []string{"event"}, WireHandlers{Event: func(context.Context, json.RawMessage) error { effects.Add(1); return nil }})
	_, _ = HandleWire(root, []string{"unrelated"}, func(context.Context, json.RawMessage) (any, error) { return 7, nil })
	_ = EmitWire(context.Background(), far, []string{"event"}, 1)
	identityWait(t, observed, bitwire.ProfileEvent)
	requireIdentityCode(t, p.Check(context.Background()), "contract_mismatch")
	requireIdentityCode(t, p.Ready(), "contract_mismatch")
	var result int
	if err = CallWire(context.Background(), far, []string{"unrelated"}, nil, &result); err != nil || result != 7 {
		t.Fatalf("shared carrier: %d, %v", result, err)
	}
	if effects.Load() != 0 {
		t.Fatal("mismatched model dispatched")
	}
	// All owned registrations were detached, permitting another interpreter.
	if _, err = RegisterWire(root, []string{"event"}, WireHandlers{Event: func(context.Context, json.RawMessage) error { return nil }}); err != nil {
		t.Fatal(err)
	}
}

func TestIdentityPreparationDefersRequestsWithoutBlockingRootAndPreservesCancel(t *testing.T) {
	near, far := localPair(t, Options{})
	installWireIdentity(t, far, preparedIdentity)
	root := testBinding(t, near)
	model := root.Select(nil)
	observed := make(chan bitwire.ProfileKind, 8)
	p, err := PrepareIdentity(observedIdentityWire{model, observed}, preparedIdentity, Options{})
	if err != nil {
		t.Fatal(err)
	}
	defer p.Close()
	var effects atomic.Int32
	_, _ = HandleWire(p.Wire(), []string{"model"}, func(context.Context, json.RawMessage) (any, error) { effects.Add(1); return 11, nil })
	_, _ = HandleWire(root, []string{"unrelated"}, func(context.Context, json.RawMessage) (any, error) { return 7, nil })
	ctx, cancel := context.WithCancel(context.Background())
	answer := make(chan error, 1)
	go func() { answer <- CallWire(ctx, far, []string{"model"}, nil, nil) }()
	identityWait(t, observed, bitwire.ProfileRequest)
	var result int
	if err = CallWire(context.Background(), far, []string{"unrelated"}, nil, &result); err != nil || result != 7 {
		t.Fatalf("root blocked: %d, %v", result, err)
	}
	cancel()
	if !errors.Is(<-answer, context.Canceled) {
		t.Fatal("request did not cancel")
	}
	identityWait(t, observed, bitwire.ProfileCancel)
	if err = p.Check(context.Background()); err != nil {
		t.Fatal(err)
	}
	if err = p.Ready(); err != nil {
		t.Fatal(err)
	}
	if err = CallWire(context.Background(), far, []string{"model"}, nil, &result); err != nil || result != 11 {
		t.Fatalf("ready model: %d, %v", result, err)
	}
	if effects.Load() != 1 {
		t.Fatalf("cancelled request dispatched: %d", effects.Load())
	}
}

func TestIdentityPreparationBoundsUnstartedAndUnboundFactories(t *testing.T) {
	for _, check := range []bool{false, true} {
		t.Run(map[bool]string{false: "unstarted", true: "unbound"}[check], func(t *testing.T) {
			near, far := localPair(t, Options{})
			installWireIdentity(t, far, preparedIdentity)
			p, err := PrepareIdentity(near, preparedIdentity, Options{RequestTimeout: 250 * time.Millisecond})
			if err != nil {
				t.Fatal(err)
			}
			defer p.Close()
			_, _ = HandleWire(p.Wire(), []string{"model"}, func(context.Context, json.RawMessage) (any, error) {
				t.Error("expired model dispatched")
				return nil, nil
			})
			if check {
				if err = p.Check(context.Background()); err != nil {
					t.Fatal(err)
				}
			}
			select {
			case <-p.released:
			case <-time.After(5 * time.Second):
				t.Fatal("preparation never expired")
			}
			if !errors.Is(p.Ready(), context.DeadlineExceeded) {
				t.Fatalf("expiry: %v", p.Ready())
			}
			binding := testBinding(t, near)
			if _, err = HandleWire(binding, []string{"model"}, func(context.Context, json.RawMessage) (any, error) { return nil, nil }); err != nil {
				t.Fatal(err)
			}
			_ = binding.Close(duplex.CodeNormal, "rebind")
			replacement, err := PrepareIdentity(near, preparedIdentity, Options{})
			if err != nil {
				t.Fatalf("identity responder leaked: %v", err)
			}
			replacement.Close()
		})
	}
}

func TestIdentityPreparationMismatchRefusesDeferredRequest(t *testing.T) {
	near, far := localPair(t, Options{})
	installWireIdentity(t, far, DeclarationIdentity{Path: "service", Digest: strings.Repeat("b", 64)})
	observed := make(chan bitwire.ProfileKind, 8)
	p, err := PrepareIdentity(observedIdentityWire{near, observed}, preparedIdentity, Options{})
	if err != nil {
		t.Fatal(err)
	}
	defer p.Close()
	_, _ = HandleWire(p.Wire(), []string{"model"}, func(context.Context, json.RawMessage) (any, error) {
		t.Error("mismatched request dispatched")
		return nil, nil
	})
	answer := make(chan error, 1)
	go func() { answer <- CallWire(context.Background(), far, []string{"model"}, nil, nil) }()
	identityWait(t, observed, bitwire.ProfileRequest)
	requireIdentityCode(t, p.Check(context.Background()), "contract_mismatch")
	requireIdentityCode(t, <-answer, "contract_mismatch")
}

func TestIdentityPreparationCancellationAndCarrierCloseReleaseDispatch(t *testing.T) {
	for _, mode := range []string{"cancel", "carrier", "close"} {
		t.Run(mode, func(t *testing.T) {
			near, far := localPair(t, Options{})
			p, err := PrepareIdentity(near, preparedIdentity, Options{})
			if err != nil {
				t.Fatal(err)
			}
			defer p.Close()
			_, _ = HandleWire(p.Wire(), []string{"model"}, func(context.Context, json.RawMessage) (any, error) {
				t.Error("abandoned model dispatched")
				return nil, nil
			})
			switch mode {
			case "cancel":
				ctx, cancel := context.WithCancel(context.Background())
				cancel()
				if !errors.Is(p.Check(ctx), context.Canceled) {
					t.Fatal("check ignored cancellation")
				}
			case "carrier":
				_ = far.Close(duplex.CodeNormal, "")
			case "close":
				p.Close()
			}
			select {
			case <-p.released:
			case <-time.After(5 * time.Second):
				t.Fatal("preparation remained blocked")
			}
			if p.Ready() == nil {
				t.Fatal("abandoned preparation became ready")
			}
		})
	}
}

func TestIdentityPreparationBoundsDeferredRequests(t *testing.T) {
	near, far := localPair(t, Options{MaxConcurrentHandlers: 8})
	installWireIdentity(t, far, preparedIdentity)
	observed := make(chan bitwire.ProfileKind, 8)
	p, err := PrepareIdentity(observedIdentityWire{near, observed}, preparedIdentity, Options{MaxConcurrentHandlers: 1})
	if err != nil {
		t.Fatal(err)
	}
	defer p.Close()
	_, _ = HandleWire(p.Wire(), []string{"model"}, func(context.Context, json.RawMessage) (any, error) { return 1, nil })
	first := make(chan error, 1)
	go func() { first <- CallWire(context.Background(), far, []string{"model"}, nil, nil) }()
	identityWait(t, observed, bitwire.ProfileRequest)
	requireIdentityCode(t, CallWire(context.Background(), far, []string{"model"}, nil, nil), "busy")
	if err = p.Check(context.Background()); err != nil {
		t.Fatal(err)
	}
	if err = p.Ready(); err != nil {
		t.Fatal(err)
	}
	if err = <-first; err != nil {
		t.Fatal(err)
	}
}
